timemachine/fe/lambda_schedule.py

```diff
@@ -19,6 +22,38 @@ def interpolate_pre_optimized_protocol(pre_optimized_protocol, num_windows):
     return lambda_schedule
 
 
+def bisect_lambda_schedule(
+    num_windows: int, lambda_interval: Optional[Tuple[float, float]] = None
+) -> NDArray[np.float64]:
+    """
+    Construct a lambda schedule by bisecting the largest difference in lambda value until
+    there are num_windows. Useful in the context of `run_sims_bisection` where states are created
+    on the fly with these same lambda values. Can save the cost of minimizing more often than necessary.
+    """
+    assert num_windows >= 2
+    # Could technically use np.linspace here for odd values, but produces minute floating point differences
+    # to the approach we use that relies on python math
+    lambda_interval = lambda_interval or (0.0, 1.0)
+    lambda_schedule_ = list(lambda_interval)
+    for i in range(num_windows - 2):
+        lambda_diffs = np.diff(lambda_schedule_)
+        # Alternate between the upper and lower section
+        # Do upper then lower to match with run_sims_bisections behavior
+        if i % 2 == 1:
+            diff_idx = np.argmax(lambda_diffs)
+            new_lamb = (lambda_schedule_[diff_idx + 1] + lambda_schedule_[diff_idx]) / 2
+            insert_idx = diff_idx + 1
+        else:
+            # Find the max from the upper end of the lambda schedule
+            diff_idx = (len(lambda_schedule_) - 1) - np.argmax(lambda_diffs[::-1])
+            new_lamb = (lambda_schedule_[diff_idx] + lambda_schedule_[diff_idx - 1]) / 2
+            insert_idx = diff_idx
+        lambda_schedule_.insert(insert_idx, new_lamb)
+
+    lambda_schedule = np.array(lambda_schedule_)
+    return lambda_schedule
```

Comment:
I think this could reuse the same logic used in the bisection on the BAR overlap objective, if we use `greedy_bisection_step` with the appropriate choice of `local_cost` and `make_intermediate`. Example:

```python
import matplotlib.pyplot as plt
import numpy as np

from timemachine.fe.protocol_refinement import greedy_bisection_step


def local_cost(lam1, lam2):
    return np.abs(lam1 - lam2)


def make_intermediate(lam1, lam2):
    return (lam1 + lam2) / 2


def step(protocol):
    protocol_, _ = greedy_bisection_step(protocol, local_cost, make_intermediate)
    return protocol_


protocol = [0.0, 1.0]

for n in range(30):
    plt.cla()
    protocol = step(protocol)
    for lam in protocol:
        plt.axvline(lam)
    plt.show()
```

Reply:
That said, would it be sufficient to just use $2^n + 1$ linearly-spaced values?

The spacings you'd get with the overlap objective for numbers of windows between these values seems sensitive to the details of the transformation, so I'm not sure they'd line up with those generated by this algorithm very well.

Reply:
> That said, would it be sufficient to just use $2^n + 1$ linearly-spaced values?
Because we do an equality check, and np.linspace produces floating point articles (0.5 not in all spacings, sometimes it is ~0.500000001 which gets treated as a different state downstream), we could also change the equality to `np.allclose` and set some threshold where we don't think there is a difference in lambda schedule.

> The spacings you'd get with the overlap objective for numbers of windows between these values seems sensitive to the details of the transformation, so I'm not sure they'd line up with those generated by this algorithm very well.

Definitely, this is not going to address all cases. But based on the benchmarks I think this is a cheap improvement that for big systems is going to make a decent wall clock improvement on.

> I think this could reuse the same logic used in the bisection on the BAR overlap objective, if we use greedy_bisection_step with the appropriate choice of local_cost and make_intermediate

Will think about how to construct it using this approach. The one difference is that this will always fill in the 'smallest' gap, which will go from the start of the array and bias towards having more evenly spaced windows. Which is why the existing code.

Such as 
[0, 1.0] - Evenly split between >= 0.5 and < 0.5
[0, 0.5, 1.0] - Odd, but odd number 
[0, 0.25, 0.5, 1.0] - Off by one
[0, 0.25, 0.5, 0.75, 1.0] - Evenly split
[0, 0.125, 0.25, 0.5, 0.75, 1.0] - off by one
[0, 0.125, 0.25, 0.375, 0.5, 0.75, 1.0] - off by two

The difference will just continue growing as the number of windows. It does shrink back down eventually, but it becomes dependent on the number of windows

Reply:
Switched to the approach you suggested in https://github.com/proteneer/timemachine/pull/1347#issuecomment-2240623648 as part of https://github.com/proteneer/timemachine/pull/1347/commits/a08b2466834d627b30d73dc3f1551b1ca0632dc8

Reply:
The implementation will probably change the benchmarks slightly in some cases, but running the vacuum legs real quickly the original states are being picked up. May make some of the benchmarks look worse since fewer windows are generated upfront, but that is more of an issue of hiding the construction time as part of the benchmark. This way we will generate strictly fewer windows than previously.